File: internal/service/usage.go
package service

import (
	"slices"

	"go.kenn.io/agentsview/internal/db"
)

type PairwiseDimension string

const (
	PairwiseDimensionModel   PairwiseDimension = "model"
	PairwiseDimensionProject PairwiseDimension = "project"
)

type PairwiseSide struct {
	Dimension PairwiseDimension `json:"dimension"`
	Value     string            `json:"value"`
	Empty     bool              `json:"empty,omitempty"`
}

type PairwiseUsageMetrics struct {
	TotalCost           float64  `json:"totalCost"`
	HasCost             bool     `json:"hasCost"`
	UnpricedModels      []string `json:"unpricedModels,omitempty"`
	InputTokens         int      `json:"inputTokens"`
	OutputTokens        int      `json:"outputTokens"`
	CacheCreationTokens int      `json:"cacheCreationTokens"`
	CacheReadTokens     int      `json:"cacheReadTokens"`
	TotalTokens         int      `json:"totalTokens"`
	SessionCount        int      `json:"sessionCount"`
	CostPerSession      *float64 `json:"costPerSession,omitempty"`
	TokensPerSession    *float64 `json:"tokensPerSession,omitempty"`
}

type PairwiseDelta struct {
	TotalCost            float64  `json:"totalCost"`
	HasCost              bool     `json:"hasCost"`
	UnpricedModels       []string `json:"unpricedModels,omitempty"`
	InputTokens          int      `json:"inputTokens"`
	OutputTokens         int      `json:"outputTokens"`
	CacheCreationTokens  int      `json:"cacheCreationTokens"`
	CacheReadTokens      int      `json:"cacheReadTokens"`
	TotalTokens          int      `json:"totalTokens"`
	SessionCount         int      `json:"sessionCount"`
	CostPerSession       *float64 `json:"costPerSession,omitempty"`
	TokensPerSession     *float64 `json:"tokensPerSession,omitempty"`
	CostRelativeChange   *float64 `json:"costRelativeChange,omitempty"`
	TokensRelativeChange *float64 `json:"tokensRelativeChange,omitempty"`
}

type PairwiseComparisonResponse struct {
	Left         PairwiseSide         `json:"left"`
	Right        PairwiseSide         `json:"right"`
	LeftMetrics  PairwiseUsageMetrics `json:"leftMetrics"`
	RightMetrics PairwiseUsageMetrics `json:"rightMetrics"`
	Deltas       PairwiseDelta        `json:"deltas"`
}

func UsageMetricsFromResult(result db.DailyUsageResult) PairwiseUsageMetrics {
	totals := result.Totals
	sessionCount := result.SessionCounts.Total
	totalTokens := totals.InputTokens + totals.OutputTokens +
		totals.CacheCreationTokens + totals.CacheReadTokens
	out := PairwiseUsageMetrics{
		TotalCost:           totals.TotalCost,
		HasCost:             totals.HasCost,
		UnpricedModels:      totals.UnpricedModels,
		InputTokens:         totals.InputTokens,
		OutputTokens:        totals.OutputTokens,
		CacheCreationTokens: totals.CacheCreationTokens,
		CacheReadTokens:     totals.CacheReadTokens,
		TotalTokens:         totalTokens,
		SessionCount:        sessionCount,
	}
	if sessionCount > 0 {
		if totals.HasCost {
			cost := totals.TotalCost / float64(sessionCount)
			out.CostPerSession = &cost
		}
		tokens := float64(totalTokens) / float64(sessionCount)
		out.TokensPerSession = &tokens
	}
	return out
}

func PairwiseDeltas(left, right PairwiseUsageMetrics) PairwiseDelta {
	out := PairwiseDelta{
		TotalCost:           right.TotalCost - left.TotalCost,
		HasCost:             left.HasCost && right.HasCost,
		InputTokens:         right.InputTokens - left.InputTokens,
		OutputTokens:        right.OutputTokens - left.OutputTokens,
		CacheCreationTokens: right.CacheCreationTokens - left.CacheCreationTokens,
		CacheReadTokens:     right.CacheReadTokens - left.CacheReadTokens,
		TotalTokens:         right.TotalTokens - left.TotalTokens,
		SessionCount:        right.SessionCount - left.SessionCount,
	}
	if !out.HasCost {
		out.UnpricedModels = sortedUniqueStrings(append(
			append([]string{}, left.UnpricedModels...), right.UnpricedModels...))
	}
	if left.CostPerSession != nil && right.CostPerSession != nil {
		v := *right.CostPerSession - *left.CostPerSession
		out.CostPerSession = &v
	}
	if left.TokensPerSession != nil && right.TokensPerSession != nil {
		v := *right.TokensPerSession - *left.TokensPerSession
		out.TokensPerSession = &v
	}
	if out.HasCost && left.TotalCost != 0 {
		v := out.TotalCost / left.TotalCost
		out.CostRelativeChange = &v
	}
	if left.TotalTokens != 0 {
		v := float64(out.TotalTokens) / float64(left.TotalTokens)
		out.TokensRelativeChange = &v
	}
	return out
}

func BuildPairwiseComparison(
	left PairwiseSide, leftResult db.DailyUsageResult,
	right PairwiseSide, rightResult db.DailyUsageResult,
) PairwiseComparisonResponse {
	leftMetrics := UsageMetricsFromResult(leftResult)
	rightMetrics := UsageMetricsFromResult(rightResult)
	return PairwiseComparisonResponse{
		Left:         left,
		Right:        right,
		LeftMetrics:  leftMetrics,
		RightMetrics: rightMetrics,
		Deltas:       PairwiseDeltas(leftMetrics, rightMetrics),
	}
}

func sortedUniqueStrings(values []string) []string {
	seen := make(map[string]bool, len(values))
	for _, v := range values {
		if v != "" {
			seen[v] = true
		}
	}
	out := make([]string, 0, len(seen))
	for v := range seen {
		out = append(out, v)
	}
	slices.Sort(out)
	return out
}

File: frontend/src/lib/api/types/usage.ts
/** Usage types — match Go structs in internal/server/usage.go
 *  and internal/db/usage.go */

export interface UsageTotals {
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalCost: number;
  hasCost?: boolean;
  unpricedModels?: string[];
}

export interface ModelBreakdown {
  modelName: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface ProjectBreakdown {
  project: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface AgentBreakdown {
  agent: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface MachineBreakdown {
  machine: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface DailyUsageEntry {
  date: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalCost: number;
  hasCost?: boolean;
  unpricedModels?: string[];
  modelsUsed: string[];
  modelBreakdowns?: ModelBreakdown[];
  projectBreakdowns?: ProjectBreakdown[];
  agentBreakdowns?: AgentBreakdown[];
  machineBreakdowns?: MachineBreakdown[];
}

export interface ProjectTotal {
  project: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface ModelTotal {
  model: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface AgentTotal {
  agent: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface MachineTotal {
  machine: string;
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  cost: number;
}

export interface CacheStats {
  cacheReadTokens: number;
  cacheCreationTokens: number;
  uncachedInputTokens: number;
  outputTokens: number;
  hitRate: number;
  savingsVsUncached: number;
}

export interface UsageSessionCounts {
  total: number;
  byProject: Record<string, number>;
  byAgent: Record<string, number>;
}

export interface UsageComparison {
  priorFrom: string;
  priorTo: string;
  priorTotalCost: number;
  deltaPct: number;
}

export interface UsageSummaryResponse {
  from: string;
  to: string;
  totals: UsageTotals;
  daily: DailyUsageEntry[];
  projectTotals: ProjectTotal[];
  modelTotals: ModelTotal[];
  agentTotals: AgentTotal[];
  machineTotals: MachineTotal[];
  sessionCounts: UsageSessionCounts;
  cacheStats: CacheStats;
  comparison?: UsageComparison;
}

export interface TopSessionEntry {
  sessionId: string;
  displayName: string;
  agent: string;
  project: string;
  startedAt: string;
  totalTokens: number;
  cost: number;
  hasCost?: boolean;
  unpricedModels?: string[];
}

export type TopUsageSessionsResponse = TopSessionEntry[];

export type PairwiseDimension = "model" | "project";

export interface PairwiseSide {
  dimension: PairwiseDimension;
  value: string;
  empty?: boolean;
}

export interface PairwiseUsageMetrics {
  totalCost: number;
  hasCost?: boolean;
  unpricedModels?: string[];
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalTokens: number;
  sessionCount: number;
  costPerSession?: number;
  tokensPerSession?: number;
}

export interface PairwiseDelta {
  totalCost: number;
  hasCost?: boolean;
  unpricedModels?: string[];
  inputTokens: number;
  outputTokens: number;
  cacheCreationTokens: number;
  cacheReadTokens: number;
  totalTokens: number;
  sessionCount: number;
  costPerSession?: number;
  tokensPerSession?: number;
  costRelativeChange?: number;
  tokensRelativeChange?: number;
}

export interface PairwiseComparisonResponse {
  left: PairwiseSide;
  right: PairwiseSide;
  leftMetrics: PairwiseUsageMetrics;
  rightMetrics: PairwiseUsageMetrics;
  deltas: PairwiseDelta;
}

export interface UsageParams {
  from?: string;
  to?: string;
  project?: string;
  machine?: string;
  agent?: string;
  model?: string;
  exclude_project?: string;
  exclude_agent?: string;
  exclude_model?: string;
  min_user_messages?: number;
  include_one_shot?: boolean;
  include_automated?: boolean;
  active_since?: string;
  timezone?: string;
}

export interface UsageTopSessionsParams extends UsageParams {
  limit?: number;
}
